### PR TITLE
Move sighash types around

### PR DESCRIPTION
**Pull request by tcharding**
This PR is now part 1 of great sighash clean up. It does not attempt to split ecdsa into two parts as discussed below, that is WIP and will be done in a separate PR after this upcoming release. Does however create `LegacySighash` and `SegwitV0Sighash` types.

This PR moves us along the road by doing:

- Move `sighash` to the `crypto` module, where I should have put it ages ago :)
- Move the taproot keys to the `keys` module
- Rename the `schnorr` module -> `taproot`
- Rename `TapSighashHash` -> `TapSighash`
- Rename `SchnorrSighashType` -> `TapSighashType`
- Fix a bunch of other schnorr usage (including pub/priv methods).

This leaves us in a situation where we have taproot sig and sighash type, and non-taproot ones prefixed with "ecdsa". Its not uniform but it kind of makes sense.

Fix: #1542
Fix: #1550

### CODE REVIEW

**Comment by tcharding** [comment]
Needs https://github.com/rust-bitcoin/rust-bitcoin/pull/1594 to get past clippy.

**Comment by Kixunil** [comment]
Yeah, I guess this is worth rebasing. :)

The following is is only important if we need to distinguish between hash-creating algos rather than signing algos. I suspect it's the latter.

I'm not sure about if it should be `schnorr::Sighash` or `taproot::Sighash`. What if a new sighash is created that still uses schnorr but not taproot? Presumably this could be the case for CISA. Also is "taproot" the name of witness v1 or of the construction? because if witness v2 uses same construction but different sighash algorithm then it'd still be confusing.

Maybe they should be called `LegacySighash` and `WitnessV1Sighash`.

**Comment by apoelstra** [comment]
@Kixunil I think we should go with `taproot::SigHash` and then deprecate it in the (far?) future when we know how it's wrong.

I definitely agree with you in principle that associating the sighash with "Taproot" is not the right categorization (though I do think it's better than "schnorr"). But the next sighash version could appear in several ways:
* a new segwit version greater than 1
* a whole new transaction format, maybe in an extension block or something
* a new tapleaf version not equal to `0xc0`
* a new CHECKSIG opcode using the same segwit/tapleaf version
* some sort of weird hybrid thing that uses the "same" sighash but adds meaningful data to the annex field, which is currently policy-limited to be empty but has no consensus limits whatsoever

The first couple are far-fetched but if I had to put odds on the what the next sighash version would look like, I'd give double-digit percentages to each of the last 3. And it would be pretty hard for us to specify all of those in the name of a type, while still making it discoverable and explainable.

Maybe when the next taproot leaf version comes out we'll deprecate much of the `taproot` module and replace it with a more specific `tapleaf_c0` module or something. But I think it's just too hard to predict this from here.

**Comment by Kixunil** [comment]
Wouldn't it be better to just not track the algorithm which *produced* the type but only track which *signing* algorithm needs to be used? Similar to what we did with `TapNode`. I can't imagine a reasonable use case where one would want to distinguish them.

Also it just occurred to me we could simply start using tagging with a default if we wanted backwards-compatibility at all costs.

**Comment by apoelstra** [comment]
I'm skeptical. I suppose that distinguishing between ECDSA and Schnorr may make certain kinds of signing APIs more type-safe, but this seems like a weird and arbitrary place to stop using the type system.

Legacy and Taproot sighashe types are generated differently, but also encoded differently, used in different kinds of outputs, and commit to different sets of transaction data. All of those differences will be true between Taproot sighashtypes and future kinds of sighashtypes.

We should probably also distinguish between legacy and segwit sighash types, because even though they have the same names, they commit to very different sets of data and have different security implications for signers.

**Comment by tcharding** [comment]
This uncovers a bit of tangle in the `taproot` code.
- taproot keys could go in `crypto/keys.rs`
- We could remove ecdsa/schnorr in favour of taproot.rs/legacy.rs/segwit_v0.rs (holds signatures, sighash, and sighash types, currently types in ecdsa would need to be duplicated for legacy/segwit_v0
- We get naming conflicts with `crate::taproot` and the re-export of `crate::crypto::taproot` (after re-name) but we could re-export the crypto stuff from `crate::taproot`.
- Some stuff in `crate::taproot` should probably be over in `crate::crypto::taproot`

I can probably do this all in two separate PRs, one for taproot stuff and one for ecdsa -> legacy/segwit_v0 if we get a bit of consensus on the direction.

**Comment by Kixunil** [comment]
@apoelstra that convinced me. I agree with also separating legacy vs segwit v0.

@tcharding can you be more specific about naming conflict in `crate::taproot`?

I know there would be a naming conflict at top-level. IME sighashes are not *that* commonly used, so I would just *not* reexport them.

Separating into two PRs sounds fine.

**Comment by tcharding** [comment]
We have `bitcoin/src/taproot.rs` and we reexport `crate::crypto::schnorr` so folks can use `schnorr::Sighash`. If that becomes `taproot::Sighash` it becomes impossible for lib users to use both modules at the same time (considering we now favour disambiguating by way of a single level of additional path). All the stuff in `src/taproot.rs` currently has `Tap` prefixes but I still do not think we should be having multiple modules with the same name without doing reexports because sooner or later it will break.

The only reason to have all the crypto stuff separate is for users to _not_ use the soon-to-be-created `crypto` crate, but for 'normal' users we can feature gate the reexport so they just use `crate::taproot` for all taproot stuff.

Anyways, I'll do a PR to showcase it.

**Comment by apoelstra** [comment]
I'd appreciate a PR. I agree with Kix that it'd be fine to just not export these types. You can only obtain them from non-generic funcitons (I'm pretty sure) so whenever users handle these they should be able to use type inference. I also think it's fine to drop `schnorr::Sighash` (or to deprecate this and require they switch to `taproot::Sighash`. I also think the various `Tap` prefixes are fine, even if you have to do ugly stuff like `taproot::TapBranch`. This is because words like "tapbranch" are standard terms by themselves.

**Comment by tcharding** [comment]
> I'd appreciate a PR.

Been working on this, thought I had a draft with some problems and while writing the PR description I gave myself some new ideas. Hence I pushed the first part to this branch, second part to come. Spoiler - the taproot stuff was easy.

**Comment by Kixunil** [comment]
> This is because words like "tapbranch" are standard terms by themselves.

Indeed, that's my thinking as well.

**Comment by tcharding** [comment]
BTW  team, lets not test against the strings in error messages, it is super painful to maintain.

**Comment by tcharding** [comment]
For anyone following along closely I got confused with the `Sighash` and `TapSighashHash` types, this PR has never touched them but I erroneously claimed yesterday to fix #1550. We now have #1620 to do that. My bad.

**Comment by tcharding** [comment]
> BTW team, lets not test against the strings in error messages, it is super painful to maintain.

EDIT: Oooh, I see why we did this now, since we no longer derive `Eq` on error types and we don't have `matches!` yet its a PITA to check a function returns the expected error type. Perhaps after MSRV bump the situation will improve. Worth keeping in mind during the error overhaul.

**Review by apoelstra — APPROVED** [review]
ACK 78a226241b047b751f69fcc37b0e9602c8165006

**Comment by Kixunil** [comment]
Nothing prevents us from writing the compiler-version-gated `matches!` macro ourselves. (similar to what we plan to do with `unsigned_abs`).

**Comment by tcharding** [comment]
Rebased to fix merge conflicts, no other changes.

**Review by apoelstra — APPROVED** [review]
ACK 30a9b3ea7a6a356b7161ffe0fd9e1e090f8e37a2

**Comment by tcharding** [comment]
Rebase only, no changes.

**Review by apoelstra — DISMISSED** [review]
ACK 72c8578526cdc5414a539e4255bf686a1fc2c01a

**Comment by tcharding** [comment]
Rebase only.

**Review by Kixunil — APPROVED** [review]
ACK be7b3754a9f62070d55b65183b09c78d145009f0

**Comment by Kixunil** [comment]
One more thing to think about: I'd like to have a type that stores sighash and sighash type and has a `sign` method which would produce `Signature`. This type would be returned from sighashcache.

Should we rename current sighashes to `*RawSighash` and name the new structs `Sighash` or do we have a better name?

**Review by apoelstra — APPROVED** [review]
ACK be7b3754a9f62070d55b65183b09c78d145009f0

File: bitcoin/src/crypto/mod.rs [hunk 1]
```diff
@@ -9,3 +9,4 @@
 pub mod ecdsa;
 pub mod key;
 pub mod schnorr;
+pub mod sighash;
```
[the remark this answers is not in the document]
Comment (apoelstra) [reply bitcoin/src/crypto/mod.rs:11]: I had the same thought but I'm not so sure it's plausible -- if you're using rust-bitcoin **on the device** and calling into secure hardware to sign, yes, but if you're running rust-bitcoin on the host, you'd provide the actual transaction data to the device and let it figure out the sighash.

In the rust-bitcoin-on-the-device case it's certainly *possible* to compute sighashes, you'll just have to copy out our code by hand. So it's not going to kill the project, though I agree it's annoying.

We can revisit this later, but for now I think the existing arrangement is reasonable.

**Comment by Kixunil** [comment]
Hmm, maybe calling the other struct `FlaggedSighash` is better?

**Comment by apoelstra** [comment]
I agree `FlaggedSighash` is better. I'd be tempted even to just make it a `type` alias (not even a newtype) around `(Sighash, SighashType)`. I think because we own both types we could still implement a `.sign` method on it, not sure.

Would also be happy with a struct with both fields `pub` if you think it's clearer to give the fields names.

We had something similar where `BitcoinSig` was a `(ecdsa::Signature, ecdsa::Sighash)` for a long time and TBH I think it worked pretty well.

This would be strictly additive so no need to get this in before the 0.30 deadline.

**Comment by Kixunil** [comment]
A struct with named `pub` fields sounds the best.

> strictly additive

Well, I want `SighashCache` to return them instead of `sighash` so it's API-breaking and also I have a use case for it. But not a big deal if it doesn't get in.

File: bitcoin/src/blockdata/transaction.rs [hunk 2]
```diff
@@ -569,16 +569,18 @@ impl<E> EncodeSigningDataResult<E> {
...
+    /// # use bitcoin::hash_types::Sighash;
     /// # let mut writer = Sighash::engine();
     /// # let input_index = 0;
     /// # let script_pubkey = bitcoin::ScriptBuf::new();
     /// # let sighash_u32 = 0u32;
     /// # const SOME_TX: &'static str = "0100000001a15d57094aa7a21a28cb20b59aab8fc7d1149a3bdbcddba9c622e4f5f6a99ece010000006c493046022100f93bb0e7d8db7bd46e40132d1f8242026e045f03a0efe71bbb8e3f475e970d790221009337cd7f1f929f00cc6ff01f03729b069a7c21b59b1736ddfee5db5946c5da8c0121033b9b137ee87d5a812d6f506efdd37f0affa7ffc310711c06c7f3e097c9447c52ffffffff0100e1f505000000001976a9140389035a9225b3839e2bbf32d826a1e222031fd888ac00000000";
     /// # let raw_tx = Vec::from_hex(SOME_TX).unwrap();
     /// # let tx: Transaction = deserialize(&raw_tx).unwrap();
-    /// if tx.encode_signing_data_to(&mut writer, input_index, &script_pubkey, sighash_u32)
+    /// # let cache = SighashCache::new(&tx);
```
[the remark this answers is not in the document]
Comment (tcharding) [reply bitcoin/src/blockdata/transaction.rs:602]: Done, https://github.com/rust-bitcoin/rust-bitcoin/pull/1653

**Comment by sanket1729** [comment]
Sorry for the late complaint here again. I will try to be an active reviewer again so that I don't fall behind again. Most of the PR is good, but I don't like the following changes:
```
Rename the schnorr module -> taproot
```
To the best of my knowledge, _everyone_ in the bitcoin ecosystem calls schnorr signatures and not taproot signatures.

I get that having all things in taproot sign API start from `Tap` is a good idea for new people who want to get started. But I think some of the terms like `schnorr::Signature` (read "Schnorr signatures") are too widespread in the existing community that using `taproot::Signature` ("taproot signatures") would just create more confusion. 

```
Rename SchnorrSighashType -> TapSighashType
```

This is something I can live with, but still have a preference for the first one.  For ecdsa case, we have `EcdsaSighashType`, but schnorr we have `TapSighashType`? I really prefer having consistent names, but this strikes as slightly odd to me.

I don't think many people outside rust-bitcoin even distinguish between the sighash types (TapSighashType vs EcdsaSighashType), so this is still a call that rust-bitcoin gets to make.

**Comment by sanket1729** [comment]
I am convinced by @Kixunil's comment here. https://github.com/rust-bitcoin/rust-bitcoin/issues/1665#issuecomment-1443136963.

Still that thrilled with `ecdsa::Signature` for pre-taproot outputs vs `taproot::Signature` for taproot outputs. But I can live with this.

**Comment by apoelstra** [comment]
>To the best of my knowledge, everyone in the bitcoin ecosystem calls schnorr signatures and not taproot signatures.

The authors of Taproot call them "BIP 340 signatures" ;).

It may be that "everyone" calls these signatures Schnorr signatures, just like they call both legacy and segwit signatures "ECDSA signatures" even though they're encoded differently and sign different things, but this sort of ambiguity is harmful when trying to create a well-typed library that doesn't let you do invalid things.

**Comment by tcharding** [comment]
Thanks for the input @sanket1729, made me think it would be kinda nice if we could push cryptography terms down into secp256k1 and in rust-bitcoin just have bitcoin terms i.e., in secp the sigs are called `schnorr::Signature` and `ecdsa::Signature` but in bitcoin they are `taproot::Signature` and `a-soon-to-be-better-named-crate-that-describes-legacy-and-segwit-v0::Signature` as we did for sighash (`LegacSighash`, `SegwitV0Sighash`, `TapSighash`)?

Its a hard line to ride between doing what everyone in the ecosystem does and doing it right (for the times when we thing there is a better way).

**Comment by tcharding** [comment]
Warning, personal opinion: I also tend to agree with the sentiment raised before that Schnorr did the wrong thing by the people when he copyrighted the signature scheme so we should not make his name a central focal point now that we can use the scheme.

**Comment by apoelstra** [comment]
In addition to disliking Schnorr, BIP-340 are literally not Schnorr signatues because they commit to the public key. They also have a specific encoding and use a particular tagged hash etc.

**Pull request merged**